opentelemetry-sdk/src/opentelemetry/sdk/trace/__init__.py

```diff
@@ -853,6 +932,11 @@ def __init__(
         self.span_processor = span_processor
         self.id_generator = id_generator
         self.instrumentation_info = instrumentation_info
+        self._limits = None
+
+    def _with_limits(self, limits: _Limits) -> "Tracer":
```

Comment:
So we're preventing from modifying the constructor of `Tracer` by using `_with_limits`, and the constructor of `Span` by using our private implementation of `_Span`? 

If we are officially recommending manually setting limits manually as a feature of OT Python, shouldn't we do things properly by just modifying the constructors (and our api surface)? Are we hesitant on doing this because it's not defined in the specs?

Reply:
> So we're preventing from modifying the constructor of Tracer by using _with_limits, and the constructor of Span by using our private implementation of _Span? And in the future, if we decide on manual configuration of limits, a new parameter into TracerProvider?

Right. I was hesitant to create a sub-class of Tracer just to enable this. I had original added the attribute directly to Span __init__ thinking it was not considered public API but then ended up moving it to _Span. I can add a similar `_with_limits()` private method to span if you think that makes more sense.

Reply:
> If we are officially supporting setting limits manually as a feature of OT Python, shouldn't we do things consistently by just modifying the constructors (and our api surface) and simply pass in the parameter down the chain to Span? Are we hesitant on doing this because it's not defined in the specs?


Yes, we should definitely do that but this PR isn't trying to implement that. I don't think spec is an issue here. The spec says it should be configurable via the TracerProvider but it leaves actual API to each language. Once we have a resolution for [this](https://github.com/open-telemetry/opentelemetry-python/discussions/1859) discussion, I'll create another PR that'll remove `_with_limits()` and add limits as a regular attribute to the constructor.

Reply:
if understood the discussion correctly, I'd prefer also having the possbility to have an individual new span to use increased limits (ie. if during design time we know and expect huge number of events etc.) but overall default limits should stay at safe numbers (to prevent bad code on other places to flood tracing)...

Hence  I'd opt for 
- `_with_limits()` at creating a new span (if the scenario described above make sense to you).
- having set the default limits set like 
`provider = TracerProvider(max_attributes_limit=100, max_events_limit=100, max_links_limit=100)` (following the [zen of Python](https://www.python.org/dev/peps/pep-0020/) where it says to keep things as easy and straight as possible from the developer perspective)

Reply:
@owais 

> Once we have a resolution for this discussion, I'll create another PR that'll remove _with_limits() and add limits as a regular attribute to the constructor.

Okay so since this [discussion](https://github.com/open-telemetry/opentelemetry-python/discussions/1859) has not been resolved yet, would it be easier to remove everything related span limit configuration through code from this PR to ge tthe "unset" feature resolved? I think some more discussion needs to be had over whether we want `_with_limits`.

Reply:
@lzchen This PR keeps everything private and does not introduce any APIs so we can change things freely depending on the outcome of the discussion. `_with_limits()` was a workaround so that I didn't have to pass down the limit values to tracer constructor as that'd change the public API. I added it specifically to unblock this PR so we can discuss the public API in more detail :)

I can remove it and set limits directly on the tracer right after creating it inside tracer provider or create a private subclass of tracer and pass limits to it's constructor but I was hoping to remove `_with_limits()` in future and replace it with the outcome of the discussion you linked. Does that make sense? LMK if you'd prefer private sub-class or directly setting the limit attrs on tracer instance instead? happy to do either.

Reply:
@SREbuilt yes, makes sense. Once we allow setting limits via tracer provider (not in this PR), you can create a dedicated tracer provider for your "special" spans and set different limits there. Any spans created with tracers fetched from that provider will respect the limits.

Reply:
@lzchen remove `_with_limits()` for now. https://github.com/open-telemetry/opentelemetry-python/pull/1839/commits/dddf654bcdb6489a83769027200f2ea452425e0e

Reply:
@owais 

Limits attr looks good for now.